proposals/0000-std-experimental.rst

```diff
@@ -0,0 +1,177 @@
+Do note that the provided approaches mix editions, channels, and feature gates.
+The goal is to illustrate that the concept of separating experimental features
+explicitly behind opt-in gates.
+
+- **Rust**: Rust has so called channels, which delineate stability by stable,
+  beta, and nightly. To opt-in locally to nightly features you'd use the
+  ``+nightly`` toolchain modifier. Effectively using a different Rust compiler
+  provided by ``rustup``.
+- **Java**: Java has the ``--enable-preview`` flag since Java 9, which allows
+  opt-in to experimental features.
+- **Gnu Compiler Collection**: GCC provides flags to opt-in to experimental
+  features through the ``-std=...`` flag. For example, it has flags like
+  ``-std=c++1z`` for experimental C++ features before they were standardized in
+  C++17.
+
+Backward Compatibility
+----------------------
+The introduction of this flag is expected to have minimal impact on existing
+code. The initial set of features behind the ``-experimental`` flag will be
+the emtpy set. Follow up proposals will be required to move existing features
+behind the `-experimental` flag. New features should by default be behind
+the `-experimental` flag, unless the respective authors consider them stable
+and not subject to sudden change.
+
+Conditionality Compilation
+--------------------------
+For maintainers there will be a ``__GHC_EXPERIMENTAL__`` macro to use with ``CPP``
+to allow for conditional compilation if ``-experimental`` is active. This can
+be used with the usual version macros to tailor to specific GHC versions as needed.
+
+Garbage Collection
+------------------
+A follow up proposal will need to address the lifecycle of features behind the
+``-experimental`` flag.  How features (not just extensions as in `#601`_),
+will either be removed or migrate from ``-experimental`` into the stable
+compiler.
+
+Relationship to ``Haskell98``, ``Haskell2010``, ``GHC2021``, and ``-fglasgow-exts``
+-----------------------------------------------------------------------------------
+GHC today provides the option to *opt-in* to a collection of Extensions
+considered stable. It does *not* permit the exclusion of experimental features.
+The proposed ``-experimental`` provides the assurances on the other end of
+the spectrum. It provides a binary option to those who want to stay out of
+highly experimental features, still allowing them to augment existing
+collections within the limits of *stable* extensions.
+
+Alternatives
+------------
+1. **Feature Flags for Each Experimental Feature**: Instead of a single flag for
+   all experimental features, individual flags for each feature could be
+   introduced. However, this could lead to a proliferation of flags and increase
+   complexity.
+2. **Separate GHC Builds**: Provide separate GHC builds for experimental
+   features. This ensures a clear separation but might be cumbersome for
+   developers to manage multiple GHC installations.
```

Comment:
Sidenote: I've been wanting to write a proposal for this. But after talking to @simonpj about his briefly, I believe it would require much more research (e.g. how does rust team deal with this), work, convincing etc.

The `-experimental` proposal seems more realistic to be accepted and implemented.